modules/fulcrum.nix

```diff
@@ -0,0 +1,136 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+let
+  options.services.fulcrum = {
+    enable = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        Enable fulcrum, an Electrum server implemented in C++.
+
+        Compared to electrs, fulcrum has a 3x larger database size but
+        can serve arbitrary address queries instantly.
+
+        This module disables peering (a fulcrum-native feature for maintaining
+        a distributed list of electrum servers), but you can manually enable
+        it via option `extraConfig`.
+      '';
+    };
+    address = mkOption {
+      type = types.str;
+      default = "127.0.0.1";
+      description = "Address to listen for RPC connections.";
+    };
+    port = mkOption {
+      type = types.port;
+      default = 50001;
+      description = "Port to listen for RPC connections.";
+    };
+    dataDir = mkOption {
+      type = types.path;
+      default = "/var/lib/fulcrum";
+      description = "The data directory for fulcrum.";
+    };
+    extraConfig = mkOption {
+      type = types.lines;
+      default = "";
+      example = ''
+        peering = true
+      '';
+      description = ''
+        Extra lines appended to the configuration file.
+
+        See all available options at
+        https://github.com/cculianu/Fulcrum/blob/master/doc/fulcrum-example-config.conf
+      '';
+    };
+    user = mkOption {
+      type = types.str;
+      default = "fulcrum";
+      description = "The user as which to run fulcrum.";
+    };
+    group = mkOption {
+      type = types.str;
+      default = cfg.user;
+      description = "The group as which to run fulcrum.";
+    };
+    tor.enforce = nbLib.tor.enforce;
+  };
+
+  cfg = config.services.fulcrum;
+  nbLib = config.nix-bitcoin.lib;
+  secretsDir = config.nix-bitcoin.secretsDir;
+  bitcoind = config.services.bitcoind;
+
+  configFile = builtins.toFile "fulcrum.conf" ''
+    datadir = ${cfg.dataDir}
+    tcp = ${cfg.address}:${toString cfg.port}
+
+    bitcoind = ${nbLib.addressWithPort bitcoind.rpc.address bitcoind.rpc.port}
+    rpcuser = ${bitcoind.rpc.users.public.name}
+
+    # Disable logging timestamps
+    ts-format = none
+
+    peering = false
+
+    ${cfg.extraConfig}
+  '';
+in {
+  inherit options;
+
+  config = mkIf cfg.enable {
+    assertions = [
+      { assertion = bitcoind.prune == 0;
+        message = "fulcrum does not support bitcoind pruning.";
+      }
+      { assertion =
+          !(config.services ? electrs)
+          || !config.services.electrs.enable
+          || config.services.electrs.port != cfg.port;
+        message = ''
+          Fulcrum and Electrs can't both bind to TCP RPC port 50001.
+          Change `services.electrs.port` or `services.fulcrum.port`
+          to a port other than 50001.
+        '';
+      }
+    ];
+
+    services.bitcoind = {
+      enable = true;
+      txindex = true;
```

Comment:
Should we warn users that enabling `fulcrum` will enable `bitcoin.txindex` which will force them to reindex their node?

Reply:
Done.